espresso/environment/4_confirmation_integrity_with_reorgs_test.go

```diff
@@ -0,0 +1,212 @@
+package environment_test
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	env "github.com/ethereum-optimism/optimism/espresso/environment"
+	"github.com/ethereum-optimism/optimism/op-e2e/e2eutils/geth"
+	"github.com/ethereum-optimism/optimism/op-e2e/system/e2esys"
+	"github.com/ethereum-optimism/optimism/op-e2e/system/helpers"
+	"github.com/ethereum-optimism/optimism/op-node/rollup/derive"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"math/big"
+	"strconv"
+	"testing"
+	"time"
+)
+
+// Computes the hash of the content of a batch. Introduced for testing purposes only.
+// @param b batch
+// @return string containing the hash of a batch
+func BatchHash(b *derive.SingularBatch) string {
+
+	// Concatenate the transactions and other relevant metadata of the batch
+	str := ""
+
+	for _, tx := range b.Transactions {
+		str = str + tx.String()
+	}
+	str = str + b.EpochHash.String()
+	str = str + strconv.Itoa(int(b.Timestamp))
+	str = str + b.ParentHash.String()
+
+	h := sha256.New()
+	h.Write([]byte(str))
+	hash := h.Sum(nil)
+
+	res := hex.EncodeToString(hash)
+
+	return res
+}
+
+// This function computes a list where at least a few batches are sent to unfinalized L1 blocks.
+// It works as follows:
+//  1. Pick the current L1 block number. Store it so that later we can reorg back to this point
+//  2. Start from the latest L2 block height.
+//  3. Pick the blocks from this height the subsequent. These blocks are unsafe (i.e. not yet sent to L1)
+//  4. Do this until the "finality" window closes i.e. before the number of new L1 blocks is bigger than  L1FinalizedDistance
+//  5. While doing this also send transactions to the sequencer
```

Comment:
What is the purpose of doing this?

Reply:
As mentioned above we want to check that in the case of a L1 reorg, all batches posted on unfinalized L1 blocks are discarded, but then posted again to L1. This is a bit tricky as one must be sure that the L2 block we look at is (1) safe which means it corresponding batch is already posted on L1 but (2) the L1 block containing this batch is not final as otherwise a L1 reorg will not affect such batch.

Reply:
Reading your question I realized that:
1. The height of the first unsafe L2 block was computed incorrectly.
2. The documentation of the function was also innacurate.

Both issues are fixed in 7f924afd8.